tools/tokenizer/src/Tokenizer.java

```diff
@@ -605,18 +608,20 @@ private static void mappingBackText(String rawFileName, String targetFileName, S
             BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(new FileOutputStream(targetFileName), "UTF8"));
 
             String buffer = "";
-            boolean isPhrase = false;
+            Double isPhrase = NOT_PHRASE;
+            final Pattern p = Pattern.compile("^<phrase_Q=([01][.][0-9]+)>$");

```

Comment:
regexp is pretty slow. is there other way to do it?

Reply:
Using only Java:
One could check if the prefix equals "<phrase_Q=" if we assume that the input is always well behaved. Is that ok?

And the double can be extracted using substring since the precision is fixed (quality scores always appears in the same interval).

I don't know any other lib/package that could be used, but I can try to search one.

Reply:
can you write a function to process character and get the numbers from the string?

Reply:
Yes. But, on a second thought, we don't really need to keep this value as a numeric value, right? Can't we just extract the substring and use it along with a bool, as before? I really don't know if this is a good approach since Java is slow when dealing with strings.

If the substring is a bad idea, I can process the string and extract the `A.B` (`<phrase_Q=A.B>`) numeric value into two ints (`A` and `B`) or a double (extra computation: `X = A + B/1000`).

Reply:
> One could check if the prefix equals "<phrase_Q=" if we assume that the input is always well behaved. Is that ok?
Just a reminder

Reply:
substring sgtm. l assume it should be faster than regular expression at least.